bindings/node/test/providers/credentialsProvider.test.js

```diff
@@ -0,0 +1,527 @@
+          expect(spiedUrl).to.not.equal(url);
+        });
+
+        it('allows custom headers to be specified', () => {
+          const options = httpSpy.args[0][1];
+          expect(options).to.have.property('headers').to.have.property('customHeader1', 'value1');
+          expect(options).to.have.property('headers').to.have.property('customHeader2', 'value2');
+        });
+      });
+    });
+
+    context('error handling', () => {
+      afterEach(() => sinon.restore());
+      context('when the request times out', () => {
+        before(() => {
+          sinon.stub(utils, 'get').rejects(new MongoCryptKMSRequestNetworkTimeoutError());
+        });
+
+        it('throws a MongoCryptKMSRequestError', async () => {
+          const error = await loadCredentials({ azure: {} }).catch(e => e);
+          expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+        });
+      });
+
+      context('when the request returns a non-200 error', () => {
+        context('when the request has no body', () => {
+          before(() => {
+            sinon.stub(utils, 'get').resolves({ status: 400 });
+          });
+
+          it('throws a MongoCryptKMSRequestError', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+            expect(error).to.match(/Malformed JSON body in GET request/);
+          });
+        });
+
+        context('when the request has a body', () => {
+          beforeEach(() => {
+            sinon
+              .stub(utils, 'get')
+              .resolves({ status: 400, body: '{ "error": "something went wrong" }' });
+          });
+
+          it('throws a MongoCryptKMSRequestError', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+          });
+
+          it('attaches the body to the error', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.have.property('body').to.deep.equal({ error: 'something went wrong' });
+          });
+        });
+      });
+
+      context('when the request returns a 200 response', () => {
+        context('when the request has no body', () => {
+          before(() => {
+            sinon.stub(utils, 'get').resolves({ status: 200 });
+          });
+
+          it('throws a MongoCryptKMSRequestError', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+            expect(error).to.match(/Malformed JSON body in GET request/);
+          });
+        });
+
+        context('when the body has no access_token', () => {
+          beforeEach(() => {
+            sinon.stub(utils, 'get').resolves({ status: 200, body: '{ }' });
```

Comment:
as a testing best practice, it's better to isolate test conditions by ensuring that the response is valid except for the thing being tested, so I would include a valid `explires_in` here

Reply:
done!